plugins/steelseries/fu-steelseries-hid.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (C) 2022 Gaël PORTAY <gael.portay@collabora.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-steelseries-common.h"
+#include "fu-steelseries-hid.h"
+
+#ifdef __USE_HID_DEVICE__
+struct _FuSteelseriesHid {
+	FuHidDevice parent_instance;
+	FuSteelseriesDeviceKind device_kind;
+};
+
+G_DEFINE_TYPE(FuSteelseriesHid, fu_steelseries_hid, FU_TYPE_HID_DEVICE)
+#else
+struct _FuSteelseriesHid {
+	FuUdevDevice parent_instance;
+	FuSteelseriesDeviceKind device_kind;
+};
+
+G_DEFINE_TYPE(FuSteelseriesHid, fu_steelseries_hid, FU_TYPE_UDEV_DEVICE)
+#endif
+
+#define STEELSERIES_BUFFER_REPORT_SIZE 64
+#define STEELSERIES_REPORT_TIMEOUT     5000
+
+#define STEELSERIES_HID_VERSION_COMMAND	       0x90U
+#define STEELSERIES_HID_VERSION_COMMAND_OFFSET 0x00U
+#define STEELSERIES_HID_VERSION_MODE_OFFSET    0x01U
+
+static gboolean
+fu_steelseries_hid_command(FuDevice *device, guint8 *data, GError **error)
+{
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(device),
+				      data[0],
+				      data,
+				      STEELSERIES_BUFFER_REPORT_SIZE,
+				      STEELSERIES_REPORT_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_NONE,
+				      error)) {
+		g_prefix_error(error, "failed to do report: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gchar *
+fu_steelseries_hid_version(FuDevice *device, GError **error)
+{
+	guint8 data[STEELSERIES_BUFFER_REPORT_SIZE] = {0};
+	const guint16 cmd = 0x90U;
+	const guint8 mode = 0U; /* string */
+
+	if (!fu_common_write_uint8_safe(data,
+					sizeof(data),
+					STEELSERIES_HID_VERSION_COMMAND_OFFSET + 1,
+					cmd,
+					error))
+		return NULL;
+
+	if (!fu_common_write_uint8_safe(data,
+					sizeof(data),
+					STEELSERIES_HID_VERSION_MODE_OFFSET + 1,
+					mode,
+					error))
+		return NULL;
+
+	if (g_getenv("FWUPD_STEELSERIES_HID_VERBOSE") != NULL)
+		fu_common_dump_raw(G_LOG_DOMAIN, "Version", data, sizeof(data));
+	if (!fu_steelseries_hid_command(device, data, error))
+		return NULL;
+	if (g_getenv("FWUPD_STEELSERIES_HID_VERBOSE") != NULL)
+		fu_common_dump_raw(G_LOG_DOMAIN, "Version", data, sizeof(data));
+
+	/* success */
+	return g_strndup((const gchar *)data, sizeof(data));
+}
+
+static gboolean
+fu_steelseries_hid_probe(FuDevice *device, GError **error)
+{
+	/* FuUsbDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_steelseries_hid_parent_class)->probe(device, error))
+		return FALSE;
+
+	/* set the physical ID */
+	return fu_udev_device_set_physical_id(FU_UDEV_DEVICE(device), "hid", error);
+}
+
+static gboolean
+fu_steelseries_hid_setup(FuDevice *device, GError **error)
+{
+	g_autoptr(GError) error_local = NULL;
+
+	g_debug("Version: '%s'\n", fu_steelseries_hid_version(device, &error_local));
```

Comment:
??

Reply:
This is debugging. I wanted you help with having an hid device.

Now, I know what the problem is; the misunderstanding of `FuUsbDevice`/`FuHidDevice` and `[USB...]`/`[HIDRAW...]`.

thanks.

Reply:
The id between all this new `GType` is to detect a device with a bluetooth connection (use hid in this case) and ask the user to switch to USB connection to get the device updated.

Is that doable?

Reply:
> ask the user to switch to USB connection to get the device updated

I'm not sure; can we discuss on the fwupd mailing list? I was asked the same thing this week by another vendor so it's something we need to think about.